### ISSUE TITLE
Training on KITTI3D

### ISSUE THREAD

**Issue by gkouros** [Angtian/NeMo#3]
Hi Angtian,

As the title suggests, I was just wondering whether you think it's possible and whether you have tried to train NeMo on KITTI3D considering that the samples contain ground truth only for azimuth and not for elevation or theta.

**Comment by Angtian** [comment]
Thanks for the suggestion. Yes, we have tried to run NeMo on KITTI3D, here is the code for convert KITTI3D annotations into our format: https://github.com/Angtian/NeMo/blob/main/code/dataset/CreateKITTI3D.py.
For the train setup, we suggest using the same hyperparameters.
Also, since the KITTI officially does not release the test set, you need to split the set into training and testing, we first split the dataset and save into ./training and ./testing

**Comment by gkouros** [comment]
Thank you for the quick response and the suggestions!

**Comment by gkouros** [comment]
I still have a few questions:

1)  I see that the script you provided, splits the training based on occlusion level into fully-visible (FV), partly-occluded (PO) and largely-occluded (LO). What is the optimal setting to train and evaluate NeMo in your opinion?
- Train on FV and evaluate on PO+LO? or
- Split all the training data into training and test sets and train on the entire training set, but evaluate separately on the FV_test, PO_test, LO_test?
Would training NeMo on occluded instances lead to worse performance?

2) I see that you use a minimum bbox threshold of 100 pixels and max truncation threshold of 0.2. Were these tuned for KITTI? Because I noticed that only 2.7k of the ~11k instances are left with those settings.

3) I used the script PredictUnlabeledDataset on KITTI with pretrained weights for PASCAL3D cars and I get really bad predictions. Do you think it's because of suboptimal recenter and resize (I used the default settings for CompCars)?

Thank you in advance.

**Comment by Angtian** [comment]
1. We split data based on occlusion level mainly for the evaluation.  To train NeMo, I would suggest using fully visible. You can split the available KITTI3D set into the training and evaluation set, and use the fully visible images in the training set to train.
Honestly, I have not tried to train on occluded data, but I think I will drop the performance a little by not a lot. Since the very occluded image only takes a small part of the whole dataset.
2. We set the threshold for minimum bbox because we think it is not very useful to estimate 3D poses for very small cars. We do have a follow-up work, which extends NeMo to estimate 6D pose, where we also evaluate small objects. But to enable such evaluation, a special training strategy is needed during backbone training. We will public the code once the paper is accepted.
3. Yes. Since the KITTI image are very different from CompCars (cars are not centered and are relatively small). For the absolute performance, you can refer to our NIPS paper (https://proceedings.neurips.cc/paper/2021/file/3a61ed715ee66c48bacf237fa7bb5289-Paper.pdf, table 2), where we apply on KITTI3D and train NeMo with 50 annotated images (with some augmentation, which gives a slight improvement when use those image without annotations). The final performance trained on the whole set will be for sure better than the 50 annos setup.

**Comment by gkouros** [comment]
Thanks again for the detailed answer. This helps a lot.

I have an additional question. I've adjusted the scripts CreateKITTI3D.py and generate_3Dpascal3D.py to generate the directories KITTI_train_NeMo and  KITTI_test_NeMo as needed for the training procedure. However, KITTI3D does not provide keypoints and their visibility like PASCAL3D and it seems that they are necessary for training NeMo. How did you do it? Did you manually annotate 50 KITTI instances or did you just use default values for keypoints and visibility?

**Comment by Angtian** [comment]
Actually, NeMo does not need the keypoints, instead, NeMo uses mesh vertex as dense points. We include and process kps in order to share code with our previous work, which mainly focuses on kps. From the code perspective, We load the kps in "[CreatePascal3DNeMo.py](https://github.com/Angtian/NeMo/blob/main/code/dataset/CreatePascal3DNeMo.py)" but replace kps with dense points in "[generate_3Dpascal3D.py](https://github.com/Angtian/NeMo/blob/main/code/dataset/generate_3Dpascal3D.py)". 
So the easiest solution is filling zeros in kps and kps visibility when creating data.

**Comment by gkouros** [comment]
Ok! This clears up my confusion. Thanks again for the quick reply!

**Comment by gkouros** [comment]
Hi again.

I have tried everything you suggested but whenever I train on KITTI3D I get really bad performance:
```
ACC_pi/6=2.9
ACC_pi/18=0.3
MedErr=85.1
```
If I evaluate on KITTI3D using a model trained on PASCAL3D I get a more reasonable performance:
```
ACC_pi/6=43.7
ACC_pi/18=68
MedErr=11.4
```

The only thing I have changed that comes to mind is the batch size to 32 instead of 108 because of limited GPU memory, however, this change had no impact when training on PASCAL3D. Do you have any idea what could be going wrong?

Thank you in advance.

**Comment by Angtian** [comment]
Hello George,
From my experience, set batch size to 32 (if you are using 4 GPU) will not have a significant impact on the final accuracy. I would suggest first checking if the vert project location is correct, which can be done by visualizing the keypoint location in the processed annotation files (like the bottom part here https://github.com/Angtian/BboxTools/blob/master/tutorial/WorkingWithImage.ipynb, but the projected vertices will be denser around the object). The second thing that might help is to check how the training loss goes, normally, it should go lower than 0.1.
